flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/data/util/DataFormatConverters.java

```diff
@@ -1515,21 +1516,40 @@ public CaseClassConverter(TupleTypeInfoBase t, DataType[] fieldTypes) {
         }
 
         @Override
-        RowData toInternalImpl(Product value) {
+        RowData toInternalImpl(Object value) {
             GenericRowData genericRow = new GenericRowData(t.getArity());
             for (int i = 0; i < t.getArity(); i++) {
-                genericRow.setField(i, converters[i].toInternal(value.productElement(i)));
+                try {
+                    genericRow.setField(
+                            i,
+                            converters[i].toInternal(
+                                    PRODUCT_CLASS
+                                            .getMethod("productElement", int.class)
+                                            .invoke(value, i)));
+                } catch (Throwable e) {
+                    // Ignore, this should never happen unless scala.Product breaks its API
```

Comment:
let's throw an exception with the cause. You never now what the future will bring.